### ISSUE TITLE
Add support for `pathlib.Path` in the `Transport` interface

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-core#5797]
Currently, that is not officially supported and there are differences between the implementations. For example `LocalTransport.isfile` accepts (not explicitly) `pathlib.Path` instances, but `SshTransport` does not. Now that `pathlib` is so common, we should provide generic and consistent support for it.

**Comment by sphuber** [comment]
Fair point, but at the very least the arguments that refer to files on the local file system should support it. We can discuss whether relative file paths that model paths on the remote should be representable by `PurePath`s or not. As you say there are pitfalls, but the current interfaec already does the same, but simply uses strings. This in a sense is worse, because a user now has to literally hard code for example the slashes in the path. The current implementations assume forward slashes, but it would be better to generalize this with `pathlib.Path`. The implementation will simply have to convert those to whatever is proper for the remote that it is modelling.

**Comment by khsrali** [comment]
We introduced `TransportPath` along with https://github.com/aiidateam/aiida-core/pull/6626
One can improve it further so it would be also support a genuine "`RemotePath`", like Chris did in `aiida-firecrest` as a quality-of-life enhancement.
